cmd/dist-receiver/main.go

```diff
@@ -0,0 +1,271 @@
+	"strings"
+	"sync"
+	"time"
+
+	b64 "encoding/base64"
+
+	config "github.com/ipfs/go-ipfs-config"
+	files "github.com/ipfs/go-ipfs-files"
+	icore "github.com/ipfs/interface-go-ipfs-core"
+	icorepath "github.com/ipfs/interface-go-ipfs-core/path"
+	ma "github.com/multiformats/go-multiaddr"
+
+	"github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core/coreapi"
+	"github.com/ipfs/go-ipfs/core/node/libp2p"
+	"github.com/ipfs/go-ipfs/plugin/loader"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+var stageNow int
+var sig chan bool
+var targets []string
+
+func main() {
+	args := os.Args
+	// 1: bootstrap node
+	// 2: cid
+	// 3: file name
+	// 4: target directory
+
+	stageNow = 0
+	sig = make(chan bool)
+
+	server := &http.Server{
+        Addr:         "0.0.0.0:4002",
+        ReadTimeout:  5 * time.Second,
+        WriteTimeout: 5 * time.Second,
+        IdleTimeout:  600 * time.Second,
+    }
+
+	http.HandleFunc("/stage", stage)
+	http.HandleFunc("/next", next)
+	http.HandleFunc("/connect", connect)
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			panic(fmt.Errorf("failed to spawn command receiver: %s", err))
+		}
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	node, _, err := spawnEphemeral(ctx, args[1])
+	if err != nil {
+		panic(fmt.Errorf("failed to spawn ephemeral node: %s", err))
+	}
+
+	if err := connectToPeers(ctx, node, []string{args[1]}); err != nil {
+		fmt.Fprintln(os.Stderr, "Failed to connect to root peer: {}", err)
+	}
+
+	stageNow = 1
+
+	<-sig
+
+	if err := connectToPeers(ctx, node, targets); err != nil {
+		fmt.Fprintln(os.Stderr, "Failed to connect to peers: {}", err)
+	}
+
+	cid := icorepath.New(args[2])
+
+	if err := node.Dht().Provide(ctx, cid); err != nil {
+		fmt.Fprintln(os.Stderr, "Failed to seed the resource file: {}", err)
+	}
+
+	if err := node.Pin().Add(ctx, cid); err != nil {
+		fmt.Fprintln(os.Stderr, "Failed to pin the resource file: {}", err)
+	}
+
+	rootNode, err := node.Unixfs().Get(ctx, cid)
+
+	if err != nil {
+		panic(fmt.Errorf("could not get file with CID: %s", err))
+	}
+
+	if err := os.RemoveAll(args[3]); err != nil {
+		panic(fmt.Errorf("could not clean previous temporary file: %s", err))
+	}
+
+	err = files.WriteTo(rootNode, args[3])
+	if err != nil {
+		panic(fmt.Errorf("could not write out the fetched CID: %s", err))
+	}
+
+	err = os.MkdirAll(args[4], os.ModePerm)
+	if err != nil {
+		panic(fmt.Errorf("failed to create target directory: %s", err))
+	}
+
+	// the following command needs variable aurgments anyway, which will cause G204 in gosec
+	cmd := exec.Command("tar", "-C", args[4], "-xzf", args[3]) // #nosec G204
+	_, err = cmd.Output()
+	if err != nil {
+		panic(fmt.Errorf("failed to uncompress resource file: %s", err))
+	}
+
+	stageNow = 2
+
+	<-sig
+
+	// the following command needs variable aurgments anyway, which will cause G204 in gosec
+	exec.Command("rm", args[3]) // #nosec G204
```

Comment:
why not use os.Remove

BTW, Is it more reasonable to put it in defer？

Reply:
Sounds like a better solution, will do